A0-1238: Marketplace for auctioning tickets

obrok opened the pull request:
# Description

Creates a new contract in `contrats/marketplace` that implements ticket auctioning.

## Type of change

Please delete options that are not relevant.

- New feature (non-breaking change which adds functionality)

# Checklist:

- I have created new documentation

deuszx commented on contracts/marketplace/lib.rs:
```diff
@@ -0,0 +1,299 @@
+        sale_multiplier: Balance,
+        ticket_token: AccountId,
+        reward_token: AccountId,
+    }
```
Since it's supposed to be an examplary contract, let's comment it generously.

obrok replied:
Are you proposing commenting these private fields? I feel like the necessary info should be exposed in the comments/docstrings on the API of the contract. I'm very much against commenting private details unless it's some unclear hack, etc.

I'm definitely pro adding a top-level docstring of some sort, describing the contract at a high level, and perhaps docstrings on the error variants, although these seem less useful, as they don't seem to be displayed anywhere anyway.

deuszx commented on contracts/marketplace/lib.rs:
```diff
@@ -0,0 +1,299 @@
+#![cfg_attr(not(feature = "std"), no_std)]
```
I'd leave a crate-level comment here explaining what the contract it is, what are the mechanics etc.

deuszx commented on contracts/marketplace/lib.rs:
```diff
@@ -0,0 +1,299 @@
+
+    #[ink(event)]
+    #[derive(Clone, Eq, PartialEq, Debug)]
+    pub struct Buy {
```
In Event Sourcing world, events would be in past tense - _bought_.

deuszx commented on contracts/marketplace/lib.rs:
```diff
@@ -0,0 +1,299 @@
+            let code_hash = Self::env()
+                .own_code_hash()
+                .expect("Failure to retrieve code hash.");
+            Initializer(code_hash)
```
Here we're calling the `Role` enum's variant directly, I'd prefer to be consistent and call it like the rest (like `admin` below)  - `Role::Initializer`.

deuszx commented on contracts/marketplace/lib.rs:
```diff
@@ -0,0 +1,299 @@
+            Role::Admin(self.this())
+        }
+
+        fn this(&self) -> AccountId {
```
This what :)

obrok replied:
It's an elixir idiom, I just didn't think twice it might not be obvious to a rustacean. I'll change it.

deuszx commented on contracts/marketplace/lib.rs:
```diff
@@ -0,0 +1,299 @@
+            }
+        }
+
+        #[ink(message)]
+        /// The length of each auction of a single ticket in blocks.
+        ///
+        /// The contract will decrease the price linearly from `average_price() * sale_multiplier()`
+        /// to `min_price()` over this period. The auction doesn't end after the period elapses -
+        /// the ticket remains available for purchase at `min_price()`.
```
I'm not sure if it's how `ink!` does it but in Rust, comments come before the macros.

deuszx commented on contracts/marketplace/lib.rs:
```diff
@@ -0,0 +1,299 @@
+        /// The multiplier applied to the average price after each sale.
+        ///
+        /// The contract tracks the average price of all sold tickets and starts off each new
+        /// auction at a price that is this multiplier times the average.
```
```suggestion
        /// auction at a price that is `self.average_price() * self.sale_multiplier()`
```

Personally, that's cleaner to me.

deuszx commented on contracts/marketplace/lib.rs:
```diff
@@ -0,0 +1,299 @@
+        #[ink(message)]
+        /// The price the contract would charge when buying at the current block.
+        pub fn price(&self) -> Balance {
+            self.current_price()
```
Since `current_price()` can be quite "expensive" and this `price()` method is available in the public API, I would try to cache the `current_price()` computation and reuse it in here.

obrok replied:
Do you mean make this `pub fn price(&mut self)`? That would make people pay gas for calling this function, wouldn't it?

obrok replied:
From your other comment, it seems you more so mean pre-computing the parts of the expression that don't depend on `current_block`?

deuszx replied:
I would probably have a field internal to the struct that is set a tuple of block height and the price computed at that height. On every call to `price` I would check if `env::block()` is equal to the cached height and if so return the cached price.

obrok replied:
Hm, that doesn't answer my question. I'm more worried when that field would be set, not how it would be computed. As far as I understand it's not a good idea to set it in `price` itself?

deuszx replied:
You would set it just where you do it now. If the check returns that `env::block() > cached.block_height` then you would recompute and update the cache.

deuszx replied:
As stated in the https://github.com/Cardinal-Cryptography/aleph-node/pull/613#issuecomment-1243959223, this optimization isn't really optimizing anything (especially that apparently ink! doesn't allow for mutable methonds in the RPC API).

deuszx commented on contracts/marketplace/lib.rs:
```diff
@@ -0,0 +1,299 @@
+                        .push_arg(from)
+                        .push_arg(self.this())
+                        .push_arg(amount)
+                        .push_arg(DUMMY_DATA),
```
Can you explain why we need this `DUMMY_DATA` arg?

fbielejec replied:
there's an argument in `psp22::transfer` that allows for passing some data along with a transfer - for messages, or callbacks

deuszx replied:
But do we need to pass it if it's unused? Can't the `psp22::transfer` make it optional?

fbielejec replied:
it's in an external crate :(

deuszx replied:
Thanks.

deuszx reviewed (commented):
I don't see any blockers. Most (if not all) my comments are more about adding documentation and code clarity. 

That being said, I think there's a benefit to caching the `current_price()` computations, as it's quite "expensive" (the parts that could be cached are `per_block_reduction`, `average_price`) and would probably be called often by other dApps/accounts to discover the price they would have to pay (although, as per my experience, if something can be computed off-chain it will be eventually b/c it's cheaper). We could also note when the price of the current ticket auction is lower than `self.min_price` and cache that the price of the ticket is `self.min_price` from now on and not recompute the `current_price` for it.

fbielejec commented on contracts/marketplace/Cargo.toml:
```diff
@@ -0,0 +1,52 @@
+scale-info = { version = "2", default-features = false, features = ["derive"], optional = true }
+
+openbrush = { git = "https://github.com/Supercolony-net/openbrush-contracts.git", rev = "8a20f95", default-features = false, features = ["psp22"] }
+access_control = { path = "../access_control", default-features = false, features = ["ink-as-dependency"] }
```
make sure all these puppies are mentioned in the std section as well

fbielejec commented on contracts/marketplace/lib.rs:
```diff
@@ -0,0 +1,299 @@
+    #[derive(Clone, Eq, PartialEq, Debug, scale::Encode, scale::Decode)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+    pub enum Error {
+        MissingRole(String),
```
why not use the actual type here? MissingRole(Role)

obrok replied:
I guess we control this type, so we could make it Encode/Decode, etc.

fbielejec replied:
it already is afaik

fbielejec commented on contracts/marketplace/lib.rs:
```diff
@@ -0,0 +1,299 @@
+    }
+
+    impl From<PSP22Error> for Error {
+        fn from(inner: PSP22Error) -> Self {
```
PSP22Error is Decode/Encode right? So you can just:
```
    impl From<PSP22Error> for Error {
        fn from(inner: PSP22Error) -> Self {
            Error::PSP22TokenCall(inner)
        }
    }
```

fbielejec commented on contracts/marketplace/lib.rs:
```diff
@@ -0,0 +1,299 @@
+        }
+
+        fn take_payment(&self, from: AccountId, amount: Balance) -> Result<(), Error> {
+            build_call::<Environment>()
```
maybe this could check the allowance before and fail with a meaningful error ?

obrok replied:
I was basically expecting the other contract to fail with a meaningful variant in that case... In `polkadot.js` all these errors look the same, where do they even show up?

fbielejec replied:
to be completely honest I'm not sure how such failure looks like, i.e. whether you get something beyond a revert, possibly not.

fbielejec commented on contracts/marketplace/lib.rs:
```diff
@@ -0,0 +1,299 @@
+
+        #[ink(message)]
+        /// The minimal price the contract allows.
+        pub fn min_price(&self) -> Balance {
```
add a setter pls

fbielejec commented on contracts/scripts/deploy.sh:
```diff
@@ -105,6 +105,43 @@ function deploy_and_instrument_game {
+
+  cargo contract call --url "$NODE" --contract "$ACCESS_CONTROL" --message grant_role --args "$AUTHORITY" 'Owner('"$contract_address"')' --suri "$AUTHORITY_SEED"
+  cargo contract call --url "$NODE" --contract "$ACCESS_CONTROL" --message grant_role --args "$AUTHORITY" 'Admin('"$contract_address"')' --suri "$AUTHORITY_SEED"
+  cargo contract call --url "$NODE" --contract "$ACCESS_CONTROL" --message grant_role --args "$game" 'Admin('"$contract_address"')' --suri "$AUTHORITY_SEED"
```
why is the game contract made an Admin?

obrok replied:
My idea here was that the game contract is going to call `reset()` after it puts in the tokens.

fbielejec reviewed (commented):
looks really good, just a few comments

fbielejec commented on contracts/marketplace/lib.rs:
```diff
@@ -0,0 +1,299 @@
+    #[ink(storage)]
+    pub struct Marketplace {
+        total_proceeds: Balance,
+        tickets_sold: Balance,
```
I know it's just an alias, but it seems weird to have Balance here, for sth that is clearly a counter - maybe use u128 explicitly

obrok replied:
On the one hand... sure. On the other the tickets are a PSP22, so a number of tickets should have type balance... I went back and forth on this, don't have an opinion, so if this doesn't convince you the other way, I'll change it.

fbielejec commented on contracts/marketplace/lib.rs:
```diff
@@ -0,0 +1,299 @@
+    pub enum Error {
+        MissingRole(String),
+        ContractCall(String),
+        PSP22TokenCall(String),
```
same here, just wrap the other error if you can

obrok replied:
I started out like that, but the other errors didn't implement something that was needed... I'll check it again on Monday.

fbielejec replied:
likely because the import wasn't in the std (there is one place where std is used, when using off-chain environment aka in unit tests

obrok commented:
> As per code organization - modelling the contract as a state machine, without any direct dependence on the other on-chain components (like calling contracts, emitting events), would make the testing much easier. There would be a component that doesn't use any on-chain functions (basically a struct with methods on it where everything "from chain" is passed in as arguments) and the outer layer that feeds expected data into it and interprets the results.

I also thought about doing it like that. However... A lot of what this contract does is interacting with the tickets and game tokens and it seemed to me like that would have to be extracted out into traits and mocked out to test the thing. That seemed like it would be way more effort than it was worth. The price calculation itself could be extracted and tested like that, but then we're only testing a small part, feels like an unsatisfying solution. What I wanted most was to instantiate all three contracts off-chain and test their interaction like that (plus use the mocking tools for managing chain state offered by ink), but that seems impossible at the moment?

obrok commented:
> That being said, I think there's a benefit to caching the current_price() computations, as it's quite "expensive" (the parts that could be cached are per_block_reduction, average_price) and would probably be called often by other dApps/accounts to discover the price they would have to pay (although, as per my experience, if something can be computed off-chain it will be eventually b/c it's cheaper).

How would one measure that something is "expensive" in this context? Since premature optimization is the root of all evil, and all.

obrok commented:
@deuszx I ran some experiments by changing the `price` function to `price(&mut self)` and comparing the transaction fee I'm charged when calling it with caching and without. If anything, the version with caching is more expensive - there's maybe a 10% difference. For now, I'm unconvinced there's any benefit to it and it complicates the code, so I'm against introducing the cache.

obrok commented:
@fbielejec @deuszx I added some docs, the `set_min_price()` message and a `max_price` to `buy()`. Please take another look, I think I've addressed everything.

deuszx commented on contracts/marketplace/lib.rs:
```diff
@@ -0,0 +1,323 @@
+        #[ink(message)]
+        pub fn buy(&mut self, max_price: Option<Balance>) -> Result<(), Error> {
+            if self.ticket_balance()? <= 0 {
+                return Err(Error::MarketplaceEmpty);
```
Maybe we should emit an event when the marketplace gets emptied? i.e. after the last ticket is sold. What do you think?

obrok replied:
It's easy enough to do, although I don't have a good feel yet of what types of things should generally be announced by events, so can't really say if it's a good idea or not.

deuszx replied:
Otherwise, how would an observer of the chain find out that the system has run out of the tickets? One would have to call `buy()` to fail with an error. That has at least two cons compared with the event:
1. Someone has to call `buy()` and pay for the txn
2. The fact marketplace is empty is returned as an error which (to me) is coincidental

I think event here (and in general) would be more explicit and visible to any observer of the contract (since that's how events work, they're "tied" to a contract) rather than watching *all* txns, filtering down to ones that call the contract and checking the result of the call.

obrok replied:
You can use `psp22::balanceOf` to check the balance of tickets available to the market. Secondly, the market exposes `ticketsAvailable`, which is basically just a wrapper around `balanceOf`.

deuszx replied:
Right, in that case I might be enough.

deuszx commented:
> > As per code organization - modelling the contract as a state machine, without any direct dependence on the other on-chain components (like calling contracts, emitting events), would make the testing much easier. There would be a component that doesn't use any on-chain functions (basically a struct with methods on it where everything "from chain" is passed in as arguments) and the outer layer that feeds expected data into it and interprets the results.
> 
> I also thought about doing it like that. However... A lot of what this contract does is interacting with the tickets and game tokens and it seemed to me like that would have to be extracted out into traits and mocked out to test the thing. That seemed like it would be way more effort than it was worth. The price calculation itself could be extracted and tested like that, but then we're only testing a small part, feels like an unsatisfying solution. What I wanted most was to instantiate all three contracts off-chain and test their interaction using the mocking tools for managing chain state offered by ink, but that seems impossible at the moment?

I can imagine abstracting away chain interaction via (simplest) a trait with methods like `call_contract`, `get_block_number`, `get_caller`, etc. (unless there's a different way to stub these) and having different implementations for on-chain use and production use where the latter would allow for registering things and then asserting them (akin a writer/state monad). That would be bigger than just for this (otherwise relatively simpler) task and could definitely use input from others when desining the API/approach.

fbielejec reviewed (approved):
lgtm

deuszx reviewed (approved):
LGTM. Ideally, there would be some tests - although apparently it's difficult to test cross-contract calls in Substrate.

I would also consider adding an event for when the marketplace is emptied.

obrok commented:
> I can imagine abstracting away chain interaction via (simplest) a trait with methods like call_contract, get_block_number, get_caller, etc. (unless there's a different way to stub these) and having different implementations for on-chain use and production use where the latter would allow for registering things and then asserting them (akin a writer/state monad). That would be bigger than just for this (otherwise relatively simpler) task and could definitely use input from others when desining the API/approach.

Yeah, I think that would be pretty useful overall. However, I'd propose doing it separately. From what I understood the plan of testing the game and related contracts is to interact with a local chain via shell scripts - parts of those scripts are already written in `contracts/scripts`.
